drud/ddev
# Track and list stopped/removed projects, fixes #642

## The Problem/Issue/Bug:

Not all projects have been visible to `ddev list` and related commands. This is a longstanding issue and affects ddev-ui, but is more relevant now that `ddev stop` actually removes containers, effectively removing ddev's knowledge of the existence of the project.

## How this PR Solves The Problem:

* Add stopped projects (which have no containers) to global config, and query from there.

## TODO:

- [x] Sort projects by active and then by name
- [x] TestMain must not remove all projects, it needs to just remove projects like "Test*"
- [x] `ddev list` should probably show only active and paused projects. Add `ddev list --all` for the rest.
- [x] Need a new `ddev stop` flag that just stops and removes from global config, doesn't do everything else (nondestructive). Use that in cmd's root_test.go setup so it's not so dangerous.
    - Projects don't get added until they're started once
    - How to delete a project from global config


## Manual Testing Instructions:

* Test functionality with ddev-ui if it's still working
* `ddev stop -a` and `ddev list` and `ddev list --all` (which will show stopped/removed projects)
* Start a few projects and `ddev list`

There are so *very* many things that can be broken with this (starting/stopping/pausing by name, whether running or not, conflicting stuff in docker from what's in global config). Please spend some serious time poking at it, and trying to make sure you can recreate bugs.

## Automated Testing Overview:

- [x] Test for consistency between docker containers and global config project list
- [x] Test for behavior when global projectlist is not fully populated (as will be the case for lots of people)
- [ ] Test for situation where there are broken or empty projects in global config

## Related Issue Link(s):

OP #642 
This should also make ddev-ui work consistently assuming it gets revived.

## Release/Deployment notes:
